### PR TITLE
[AutoDiff] Rewrite `AdjointEmitter`’s memory management.

### PR DESCRIPTION
**Pull request by rxwei**
- Make `AdjointValue` carry values with cleanup (`ValueWithCleanup`), and accumulate them to form a dag. For pullbacks’ return values, disable the top-level cleanup and recursively apply all cleanups.
- Make `AdjointValue` materializers in `AdjointEmitter` take rvalue references, since each `AdjointValue` is supposed to be materialized at most once.
- No longer “differentiate” reference counting instructions since they are easier managed separately.
- Separate lvalues (“adjoint buffers”) from rvalues (“adjoint values”).
- Use `AdditiveArithmetic.+=` for adjoint accumulation into buffers.

Resolves current memory leaks/hacks in AD.

### CODE REVIEW

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -187,23 +187,18 @@ static CanType joinElementTypesFromValues(SILValueRange &&range,
...
-                                                       ProtocolDecl *protocol) {
+/// Given an operator name, such as '+', and a protocol, returns the '+'
+/// operator. If the operator does not exist in the protocol, returns null.
+static FuncDecl *findOperatorDeclInProtocol(DeclName operatorName,
+                                            ProtocolDecl *protocol) {
   assert(operatorName.isOperator());
   // Find the operator requirement in the `VectorNumeric` protocol
   // declaration and cache it.
   auto opLookUp = protocol->lookupDirect(operatorName);
   // Find the `+` with type siguature `(Self, Self) -> Self`.
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]:
```suggestion
  // Find the `+` with type signature `(Self, Self) -> Self`.
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 2]
```diff
@@ -2611,107 +2621,242 @@ bool AdjointGen::run() {
...
 namespace {
 class AdjointEmitter;
+class AdjointValue;
 
-/// A symbolic adjoint value that is capable of representing zero value 0 and
-/// 1, in addition to a materialized SILValue. This is expected to be passed
-/// around by value in most cases, as it's two words long.
-class AdjointValue {
-  friend class AdjointEmitter;
+class Cleanup {
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Could you add a doc comment for `Cleanup` and `ValueWithCleanup` explaining their purpose/usage?

**Review by dan-zheng — APPROVED** [review]

**Comment by jekbradbury** [comment]
For the CIFAR convnet code (or a reduced, no-`var` version that's really just a single reshape + dense), this branch either gives an assertion failure in
```
-> 512          assert(getStrongExtraRefCount() >= dec  &&
   513                 "releasing reference whose refcount is already zero");
```
or segfaults elsewhere (including, strangely, in the code to print the assertion message). Here's some SIL that might be helpful: https://gist.github.com/jekbradbury/ce4598ea56265fc730a057d66a0c531c

**Comment by jekbradbury** [comment]
Here's a minimal reproducer for the SwiftShims `Assertion `getStrongExtraRefCount() >= dec && "releasing reference whose refcount is already zero"' failed.`:
```
import TensorFlow

let x = Tensor<Float>(zeros: [1, 1])

let w = Tensor<Float>(zeros: [1, 1])
let b = Tensor<Float>(zeros: [1])

var model = Dense<Float>(inputSize: 1, outputSize: 1, activation: {$0})

for _ in 0..<2 {
    let _ = gradient(at: model) { model in
        // (matmul(x, w) + b).sum()
        model.applied(to: x).sum()
    }
}
```
Note that if I use the commented-out line instead it runs without error. Question: did you run the swift-apis tests on this branch?

**Comment by jekbradbury** [comment]
Potentially more targeted repro:
```swift
import TensorFlow

struct Foo<Scalar>: Differentiable
    where Scalar: FloatingPoint & Differentiable & TensorFlowScalar {

    var bar: Tensor<Scalar>
    var baz: Tensor<Scalar>

    @differentiable(wrt: (self, input))
    func applied(to input: Tensor<Scalar>) -> Tensor<Scalar> {
        return self.bar + input
    }
}

let x = Tensor<Float>(zeros: [1])
var model = Foo<Float>(bar: x, baz: x)

for _ in 0..<2 {
    let _ = gradient(at: model) { model in
        model.applied(to: x).sum()
    }
}
```

**Comment by rxwei** [comment]
Thanks. Debugging this.

**Comment by rxwei** [comment]
I know what's wrong. We are missing a retain after we emit a `struct` instruction.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@jekbradbury Fixed. Please verify.

**Comment by jekbradbury** [comment]
Yes, this is vastly better and should be sufficient for now! With side effecting code, it still appears to leak a little bit (my guess is we couldn't train an ImageNet model to convergence) but it's almost completely leak-free on code with `let`s.

**Comment by rxwei** [comment]
If you send me the canonical SIL of the model that uses a `var` to chain layers, i should be able to figure out what's wrong with side effects.

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by jekbradbury** [comment]
Model code:
```swift
public struct PyTorchModel : Layer {
    var conv1: Conv2D<Float>
    var pool: MaxPool2D<Float>
    var conv2: Conv2D<Float>
    var dense1: Dense<Float>
    var dense2: Dense<Float>
    var dense3: Dense<Float>
    public init() {
        conv1 = Conv2D<Float>(filterShape: (5, 5, 3, 6), padding: .valid)
        pool = MaxPool2D<Float>(
            poolSize: (2, 2), strides: (2, 2), padding: .valid)
        conv2 = Conv2D<Float>(filterShape: (5, 5, 6, 16), padding: .valid)
        dense1 = Dense<Float>(
            inputSize: 16 * 5 * 5, outputSize: 120, activation: relu)
        dense2 = Dense<Float>(inputSize: 120, outputSize: 84, activation: relu)
        dense3 = Dense<Float>(inputSize: 84, outputSize: 10, activation: { $0 })
    }
    @differentiable(wrt: (self, input))
    public func applied(to input: Tensor<Float>) -> Tensor<Float> {
        var tmp = input
        tmp = pool.applied(to: relu(conv1.applied(to: tmp)))
        tmp = pool.applied(to: relu(conv2.applied(to: tmp)))
        let batchSize = tmp.shape[0]
        tmp = tmp.reshaped(
            toShape: Tensor<Int32>([batchSize, Int32(16 * 5 * 5)]))
        tmp = dense1.applied(to: tmp)
        tmp = dense2.applied(to: tmp)
        tmp = dense3.applied(to: tmp)
        return tmp
    }
}
```
Canonical SIL: https://gist.githubusercontent.com/jekbradbury/ce4598ea56265fc730a057d66a0c531c/raw/5f6118dfdaa037c3750a772e8a3d5abeaeca77d4/vars.sil

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@jekbradbury Found the issue. Fixing coming.

**Pull request merged**